promql-compliance-tester.yml

```diff
@@ -30,6 +35,20 @@ query_tweaks:
   # UNCOMMENT FOR VICTORIAMETRICS:
   #- note: 'VictoriaMetrics incorrectly modifies incoming start/end range query timestamps to be aligned to the resolution step size.'
   #  align_timestamps_to_step: true
+  #
+  # UNCOMMENT FOR NEW RELIC:
+  # - note: 'New Relic is sometimes off by 1ms when parsing floating point start/end timestamps.'
+  #   truncate_timestamps_to_ms: 1000
+  # - note: 'New Relic adds a "prometheus_server" label to distinguish Prometheus servers, leading to extra labels in query results. These need to be stripped before comparisons.'
+  # - note: 'New Relic does not consistently include the "__name__" attribute in its responses.'
+  #   no_bug: true
+  #   drop_result_labels:
+  #     - prometheus_server
+  # - note: 'New Relic omits the first resolution step in the output.'
+  #   ignore_first_step: true
+  # - note: 'New Relic stores metric counters as deltas, not cumulative values. New Relic will recognize these as guages by if "_cnt" is appended to the metric name.'
+  #   query_string_replacements:
+  #     _total: _total_cnt
```

Comment:
In the example test query configs for New Relic, I only replaced `_total` in some cases, namely where a counter metric was used in a way where we would look at its absolute value. I kept the normal `_total` suffix in queries that looked at the relative increase, e.g. by running `rate()` / `irate()` etc. on the counter metric. As far as I understand, this directive will just convert those latter cases to gauges as well, and then I'm not sure what NR's implementation of `rate()` etc. will yield when applied to gauges?

Reply:
I hadn't noticed that.  I will review the items that didn't append "_cnt" in your original customized NR config.

Reply:
I believe those relative increase methods only work with the guage type, which is what the _cnt extension causes the values to be interpreted as on the NR side.  

If I run the tester with just the relative increase queries, appending `_cnt` increases the success rate from 0% to 8.11%.  (Remaining failures, I believe, are unrelated to the `_cnt` extension.

Reply:
> I believe those relative increase methods only work with the guage type

This would strike me as odd, as those functions (`rate()` / `irate()` / `increase()` in Prometheus) are only meant to be used on counters, not on gauges (for those we have `deriv()`, `delta()`, etc.).

When I manually test `rate(demo_api_request_duration_seconds_count[1m])` vs. `rate(demo_api_request_duration_seconds_count_cnt[1m])`, the counter variant is also the one that looks closest to the native Prometheus:

**With counter metric in Prometheus (same for gauge, since Prometheus doesn't know the type, only looks at the value):**

![rate-on-prom-counter](https://user-images.githubusercontent.com/538008/96738966-938b6700-13bf-11eb-9f04-f725410a50cd.png)

**With counter-interpreted metric in New Relic:**

![rate_on_nr_counter](https://user-images.githubusercontent.com/538008/96738970-94bc9400-13bf-11eb-8a1e-3ad77f6976e7.png)

**With gauge-interpreted metric in New Relic:**

![rate_on_nr_gauge](https://user-images.githubusercontent.com/538008/96738972-94bc9400-13bf-11eb-9189-6b3f0590273b.png)

So we definitely want to interpret a counter-style metric as a New Relic counter when we plug it into `rate()` and friends.

Reply:
> If I run the tester with just the relative increase queries, appending _cnt increases the success rate from 0% to 8.11%.

Following up on this as well, I wonder whether we are talking about the same test cases. The relevant test case for `rate`/`irate`/`increase` (all the counter-related increase functions) would be this one:

```yaml
  - query: '{{.extrapolatedRateFunc}}(demo_cpu_usage_seconds_total[{{.range}}])'
    variant_args: ['extrapolatedRateFunc', 'range']
```

But I get a 0% passing rate for that, no matter if I turn on the query replacement query tweak or manually append `_cnt` or just keep it as it is. (Still, as seen above, keeping it a counter-interpreted metric for those functions leads to a visually closer result to Prometheus)

Reply:
(I am running this against my personal NR account by the way, so not sure if you're already testing with some account that has more features or fixes enabled).

Reply:
@juliusv @ccobb-nr I've added an "exemptions" to the yaml, where functions can be added to exempt them from the metric name substitution. Addressed in [8896bb41961dff0d4a53997bb87f4bc13a9157a1](https://github.com/promlabs/promql-compliance-tester/pull/2/commits/8896bb41961dff0d4a53997bb87f4bc13a9157a1).